crates/bevy_animation/src/curve_path.rs

```diff
@@ -0,0 +1,159 @@
+//! A curve through space.
+
+use bevy_math::cubic_splines::{CubicCurve, Point};
+
+/// A curve representing a path through space. Unlike [`CubicCurve`], it is sampled based on distance
+/// along the path. Can be used for linearly animating along a curve.
+///
+/// Note: Values are approximated using a series of line segments. Accuracy is based on the
+/// subdivisions specified when creating the path.
+#[derive(Clone, Debug)]
+pub struct CurvePath<P: Point> {
+    curve: CubicCurve<P>,
+    arc_lengths: Vec<f32>,
+}
+
+impl<P: Point> CurvePath<P> {
+    /// Creates a new [`CurvePath`] from a curve. Subdivisions will determine how many line segments
+    /// are used, and therefore the accuracy.
+    pub fn from_cubic_curve(curve: CubicCurve<P>, subdivisions: usize) -> CurvePath<P> {
+        let arc_lengths: Vec<f32> = curve
+            .iter_positions(subdivisions)
+            .scan((0.0, curve.position(0.0)), |state, x| {
+                state.0 += x.distance(state.1);
+                state.1 = x;
+
+                Some(state.0)
+            })
+            .collect();
+
+        CurvePath { curve, arc_lengths }
+    }
+
+    /// Returns the [`CubicCurve`] this path is based on.
+    pub fn curve(&self) -> &CubicCurve<P> {
+        &self.curve
+    }
+
+    /// Total length of the curve.
+    pub fn length(&self) -> f32 {
+        self.arc_lengths[self.arc_lengths.len() - 1]
+    }
+
+    /// Returns a 't' value corresponding to the `t` value on the underlying curve that matches the
+    /// provided `distance`.
+    pub fn t(&self, distance: f32) -> f32 {
+        // Get index with greatest value that is less than or equal to target distance.
+        let closest = self.arc_lengths.partition_point(|&x| x <= distance) - 1;
+
+        // Check if index's distance perfectly matches target distance, otherwise lerp between the
+        // index and its next neighbor.
+        if self.arc_lengths[closest] == distance {
+            (closest + 1) as f32 / (self.arc_lengths.len() + 1) as f32
```

Comment:
What is the purpose of the `+1` in `self.arc_lengths.len() + 1`?
Seems incompatible with line 60, where one divides by `self.arc_lengths.len() - 1` instead.
I think at least one of these is incorect then.

Also what is the reason for the `+1` in `closest + 1`, if `self.arc_lengths[closest] == distance`, then shouldn't we be returning the `t` value corresponding to `closest` and not `closest + 1`?